DOC, TST: enable doctests

tylerjereddy opened the pull request:
Related to #9415, ~but using doctests proper--this is pretty painful, although I'm making progress & there are some real issues in the docstrings.~

Using this command locally: `python runtests.py -- --doctest-modules --ignore=numpy/ma/version.py --ignore=numpy/f2py/__main__.py --ignore=numpy/distutils/msvc9compiler.py --ignore=numpy/core/generate_numpy_api.py --ignore=numpy/core/cversions.py --ignore=numpy/testing -n 3 >& log.txt`

To do:

- [x] resolve remaining doctest errors locally
- [x] hook up to CI with a 3.x doctest job

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -2020,9 +2019,10 @@ def luf(lamdaexpr, *args, **kwargs):
     >>> np.ndarray(shape=(2,2), dtype=float, order='F')
-    array([[ -1.13698227e+002,   4.25087011e-303],
-           [  2.88528414e-306,   3.27025015e-309]])         #random
+    array([[0.0e+000, 0.0e+000],
```
Seeding isn't going to help - this is uninitialized memory

ev-br replied:
One other advantage of a custom doctester: can recognize a comment-like human-readable marker, `# random` or `# may vary`.

tylerjereddy replied:
Yeah, I switched back to that approach now that we use refguide. Technically NumPy had its own checker but I guess it is long dead now vs. the SciPy one I've ported over.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -2818,18 +2818,17 @@ def luf(lamdaexpr, *args, **kwargs):
     >>> A = np.array([1, 256, 8755], dtype=np.int16)
     >>> map(hex, A)
-    ['0x1', '0x100', '0x2233']
+    <map object at 0x...>
```
I don't think this was the intent of the test - add a `list` on the line above instead. Same below

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -4163,11 +4166,11 @@ def luf(lamdaexpr, *args, **kwargs):
     True
     >>> x.tobytes('F')
-    b'\\x00\\x00\\x00\\x00\\x02\\x00\\x00\\x00\\x01\\x00\\x00\\x00\\x03\\x00\\x00\\x00'
+    b'\\x00\\x00\\...'
```
Again, this defeats the purpose of the test. I think the right fix here is to pass `dtype='<u2` above - it's probably failing due to the platform-specific nature of `np.dtype(int)`

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -4356,9 +4359,8 @@ def luf(lamdaexpr, *args, **kwargs):
     >>> y.view(dtype=[('width', np.int16), ('length', np.int16)])
-    Traceback (most recent call last):
-      File "<stdin>", line 1, in <module>
-    ValueError: new type not compatible with array.
```
Showing the message is useful here

tylerjereddy replied:
Shouldn't be any cases left where I've obliterated Tracebacks anymore -- if there are, that's probably because they don't happen in execution anymore for whatever reason.

tylerjereddy replied:
Double checked the source -- the modernized traceback is in place here now so marking as resolved.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -5535,7 +5537,7 @@ def luf(lamdaexpr, *args, **kwargs):
     >>> np.negative.at(a, [0, 1])
     >>> print(a)
-    array([-1, -2, 3, 4])
+    [-1 -2 3 4]
```
Better to remove the print on the line above

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -5616,12 +5618,12 @@ def luf(lamdaexpr, *args, **kwargs):
 
     >>> np.dtype([('f1', np.uint), ('f2', np.int32)])
-    dtype([('f1', '<u4'), ('f2', '<i4')])
+    dtype([('f1', '<u8'), ('f2', '<i4')])
```
This isn't going to help if we switch back to linux. Use a sized type instead.

tylerjereddy replied:
I've tried to enforce sized types where I observed issues like this throughout.

tylerjereddy replied:
I switched to a sized unsigned integer type so marking as resolved

eric-wieser commented on numpy/core/numeric.py:
```diff
@@ -376,13 +376,13 @@ def full_like(a, fill_value, dtype=None, order='K', subok=True):
+    array([0.1, 0.1, 0.1, 0.1, 0.1, 0.1])
     >>> np.full_like(x, np.nan, dtype=np.double)
-    array([ nan,  nan,  nan,  nan,  nan,  nan])
+    array([nan,  nan,  nan,  nan,  nan,  nan])
```
This spacing looks wrong to me. I thought we killed these extra spaces

tylerjereddy replied:
That's what the code produces

tylerjereddy replied:
I'll remove the intervening spaces locally -- the important part of the diff was removing the leading space, but that may have been left over from the doctest-strict changes. The intervening spaces are probably just fine for refguide so that passes through

tylerjereddy replied:
pushed the space clean-up, marking as resolved

eric-wieser commented:
I'm not sure I like the use of ellipsis for floats in the docstrings, as they make the reprs non-eval-able - although I don't know what the right answer there is.

Lots of other good stuff here, thanks!

rgommers commented:
> I'm not sure I like the use of ellipsis for floats in the docstrings, as they make the reprs non-eval-able - although I don't know what the right answer there is.

I agree, that doesn't look quite right. Using the refguide checker as discussed in gh-9415 should fix that, it doesn't require exact repr equality but just that numerical values are within a given tolerance (uses `assert_allclose` under the hood IIRC). Because of this type of thing, I think that refguide checker is strongly preferred over using `doctest`.

> Lots of other good stuff here, thanks!

+1

tylerjereddy commented:
I also wonder about just creating a lightweight [doctest flag](https://docs.python.org/3.7/library/doctest.html#doctest.register_optionflag) for the floating point comparisons rather than copying over the entire `refguide_check.py` module from SciPy. The process, I think, is to subclass `OutputChecker` from the standard library, which [refguide does too](https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L514), and use [`np.allclose` as in refguide_check.py](https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L623) as mentioned above.

I guess one issue here is that NumPy can't depend on SciPy, so it would almost be better if the refguide checker base code lived in NumPy instead, and perhaps even better still if there were a general scientific refguide checker maintained externally to both for use in the ecosystem. It is a little ugly to have two copies of `refguide_check.py` in both NumPy and SciPy, although time investment for alternatives may be higher than the cost of the sync / maintenance burden I suppose.

shoyer commented on numpy/core/_add_newdocs.py:
```diff
@@ -320,43 +320,42 @@
-                for (a, b, c) in it:
-                    addop(a, b, out=c)
-            return it.operands[2]
+        >>> def iter_add_py(x, y, out=None):
```
I'm not sure these added indentation markers are an improvement in readability... it would be great if doctests could execute multiline function definitions with it.

tylerjereddy replied:
Would be nice -- maybe if / when we generate an ecosystem "scientific docstrings" package of sorts.

tylerjereddy commented:
Also worth noting that `add_newdocs.py` is sub-optimal for doctest error / trace reporting because of the source / doc dissonance. Historically it seems that the project had some custom doctest directives like `#random`, but the unit tests for those haven't been running it seems.

Indeed, we also have / have had our own custom ouput checkers, subclassed from `doctest`, living in `numpy/testing/_private/noseclasses.py` and `NumpyDoctest` calls nosetests. There's probably some complicated history there.

ev-br commented:
Agreed a small package living in the numpy org would be best. Allowing for per-project skip lists etc does not souns difficult. Whether the package can be used as an external dependency or will it need to be vendored (like numpydoc in scipy) remains to be seen --- I guess one needs to try adapring it foe numpy and see :--).

tylerjereddy commented:
Now working with a crudely-adjusted version of SciPy's `refguide_check` module & working through the plethora of issues there. I'll have to backtrack a bit to remove my previously-introduced ellipses, so still quite some way from being ready for review.

Added the refguide check tentatively to Azure CI in this branch -- it sould fail with around 4400 lines of problematic output (it was 6000+ earlier today).

One thing I don't get is the magic behind printing array objects resulting in the output of `random` -- it seems I always need to replace the `print(x)` with simply `x` to display an array in doctest context.

tylerjereddy commented on numpy/core/arrayprint.py:
```diff
@@ -1582,8 +1584,8 @@ def set_string_function(f, repr=True):
     HA! - What are you going to do now?
-    >>> print(a)
-    [0 1 2 3 4 5 6 7 8 9]
+    >>> _ = a
```
`print()` will produce `random` here and elsewhere

ev-br commented on numpy/lib/_datasource.py:
```diff
@@ -646,15 +648,15 @@ class Repository (DataSource):
-        ...     fp = repos.open(filename)
-        ...     fp.analyze()
-        ...     fp.close()
+        >>> repos = np.lib._datasource.Repository('/home/user/data/dir/') # doctest: +SKIP
```
In scipy, we skip pseudocode in two ways: there's a list of functions which are not supposed to be doctested outright, https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L109, and there's a list of pseudocode markers, https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L786

tylerjereddy replied:
I'll mark this as resolved in the sense that I'm now using those lists to avoid repeated inline use of skip directives.

tylerjereddy commented:
- `refguide_check` is now passing in CI for this branch, so that's a good sign
- I went through and tried to manually remove all cases where ellipses were being used for floating point comparisons (ouch), before we switched to refguide and random directives
- of course more clean-up is needed with this big of a diff, but this should be looking much more respectable now

tylerjereddy committed:
TST, DOC: enable refguide_check

* ported the refguide_check module from SciPy for usage
in NumPy docstring execution/ verification; added the
refguide_check run to Azure Mac OS CI

* adjusted NumPy docstrings such that refguide_check passes

tylerjereddy commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -1558,33 +1559,38 @@ def add_newdoc(place, name, doc):
     array([242], dtype=uint8)
-    >>> np.binary_repr(x, width=8)
+    >>> # currently raises TypeError:
+    >>> np.binary_repr(x, width=8) # doctest: +SKIP
```
For this case and the one below, I could use some advice on how we'd like to refactor this docstring given that these commands raise exceptions in the master branch.

eric-wieser replied:
Just remove the `[]` in the array above and create a 0d array, then the function will work

tylerjereddy replied:
That's helpful, thanks

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -294,8 +294,9 @@ def printoptions(*args, **kwargs):
     >>> with np.printoptions(precision=2):
-    ...     print(np.array([2.0])) / 3
+    ...     np.array([2.0]) / 3
     [0.67]
```
This surprises me - shouldn't it be showing the `repr` here?

tylerjereddy replied:
The true result is `array([0.67])` if that's what you mean, but refguide_check has a number of fallback switches and seems to be "smart enough" to pass with just `[0.67]`.

I'm changing to the more formally correct result as you imply here, but probably won't go hunting for these cases too actively for now since some leniency on the "printed" values is one of the reasons for picking refguide over vanilla doctest.

ev-br replied:
@eric-wieser it basically does `assert_allclose(eval(want), eval(got))`

eric-wieser commented on numpy/ma/extras.py:
```diff
@@ -1524,20 +1545,20 @@ def flatnotmasked_edges(a):
-    >>> print(flatnotmasked_edges(ma))
-    None
+    >>> np.ma.flatnotmasked_edges(a)
+    >>> # would print None
```
Why not leave the `print`?

tylerjereddy replied:
True -- that works now. `print` sometimes causes issues because it will generate an expected value of `random` when the argument is an array

eric-wieser commented on numpy/ma/extras.py:
```diff
@@ -1078,12 +1097,12 @@ def intersect1d(ar1, ar2, assume_unique=False):
-           fill_value = 999999)
+    >>> x = np.ma.array([1, 3, 3, 3], mask=[0, 0, 0, 1])
+    >>> y = np.ma.array([3, 1, 1, 1], mask=[0, 0, 0, 1])
+    >>> np.intersect1d(x, y)
```
This line should be calling `np.ma.intersect1d`

eric-wieser commented on numpy/ma/extras.py:
```diff
@@ -492,28 +491,47 @@ def apply_over_axes(func, a, axes):
+      mask=[[[False,  True, False, False],
+             [ True,  True,  True,  True],
+             [False, False, False, False]],
+    <BLANKLINE>
```
Does refguide provide a way to avoid needing this?

tylerjereddy replied:
Yup, I can just delete `<BLANKLINE>` apparently

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -2583,18 +2583,21 @@ def multi_dot(arrays): return functools.reduce(np.dot, arrays)
     >>> # the actual dot multiplication
-    >>> multi_dot([A, B, C, D])
+    >>> _ = multi_dot([A, B, C, D])
+    ...
```
I don't think you need both the `...` and the `_ =` - either should work alone

tylerjereddy replied:
True

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -992,27 +988,25 @@ def dsplit(ary, indices_or_sections):
+    [array([[[ 0.,  1.],
+            [ 4.,  5.]],
+           [[ 8.,  9.],
+            [12., 13.]]]), array([[[ 2.,  3.],
```
Why did you need to rewrap this line but not the ones below?

tylerjereddy replied:
This is what NumPy produces on master branch locally for me -- as noted above this PR was initially aimed at getting doctest to pass and then relaxed to just require passing refguide, so the superset of changes in the PR probably exceeds the absolute minimum requirement for refguide to pass.

I'll mark this particular one as resolved since the change matches master branch output formatting.

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -992,27 +988,25 @@ def dsplit(ary, indices_or_sections):
             [  4.,   5.,   6.,   7.]],
            [[  8.,   9.,  10.,  11.],
-            [ 12.,  13.,  14.,  15.]]])
+            [12.,  13.,  14.,  15.]]])
```
I think you need to touch the lines above too

tylerjereddy replied:
True, although refguide lets it through regardless. I'll do that now

tylerjereddy commented on numpy/lib/function_base.py:
```diff
@@ -2311,11 +2310,15 @@ def cov(m, y=None, rowvar=True, bias=False, ddof=None, fweights=None,
-        >>> cov = np.dot(m * w, m.T) * v1 / (v1**2 - ddof * v2)
+        >>> # NOTE: line below highly problematic in current NumPy
+        >>> m -= np.sum(m * w, axis=1, keepdims=True) / v1 # doctest: +SKIP
+        >>> cov = np.dot(m * w, m.T) * v1 / (v1**2 - ddof * v2) # doctest: +SKIP
```
Maybe someone familiar with the covariance code has suggestions for the failing docstring cases above, currently marked with SKIP directives?

tylerjereddy replied:
Adjusted the example to produce the same result as `np.cov()` with the same input parameters. I had to define the degrees of freedom as that variable was missing in the code block.

Marking as resolved.

tylerjereddy commented:
I've lifted the WIP tag now. That doesn't mean the changes are immaculate, but I have gone through the full diff to fix anything I thought was blatantly wrong, or being skipped when it could be fixed easily.

The superset of changes is probably slightly larger than what refguide formally requires to pass because I started out using doctest proper and then reverted, but I should have reeled most of the excess in now.

eric-wieser commented on numpy/core/defchararray.py:
```diff
@@ -1084,23 +1079,20 @@ def lstrip(a, chars=None):
-        dtype='|S7')
+    array(['aAaAaA', '  aA  ', 'abBABba'], dtype='<U7')
     >>> (np.char.lstrip(c, ' ') == np.char.lstrip(c, '')).all()
     ... # XXX: is this a regression? this line now returns False
```
Change to "This used to return True" to reflect your change here

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -763,14 +764,15 @@ def einsum_path(*operands, **kwargs):
     >>> print(path_info[0])
     ['einsum_path', (1, 2), (0, 1)]
     >>> print(path_info[1])
-      Complete contraction:  ij,jk,kl->il
```
This seems suspiscious to me

tylerjereddy replied:
The contraction is indeed constant in this case, but the FLOP counts don't seem to be fully reproducible or at least fully portable, which is maybe not surprising? The `# may vary` designation may seem to instill a lack of confidence in the contraction, but it was probably intended for the FLOP count variability (the entire block is labelled as variable as a result).

charris commented on numpy/core/numerictypes.py:
```diff
@@ -163,19 +163,19 @@ def maximum_sctype(t):
+    <class 'numpy.uint64'>
     >>> np.maximum_sctype(complex)
-    <type 'numpy.complex192'>
+    <class 'numpy.complex256'>
```
Needs `# may vary`. In general, `complex256` may vary.

eric-wieser replied:
Probably needs `# may vary`

charris commented on numpy/core/numerictypes.py:
```diff
@@ -163,19 +163,19 @@ def maximum_sctype(t):
+    <class 'numpy.int64'>
     >>> np.maximum_sctype('f4')
-    <type 'numpy.float96'>
+    <class 'numpy.float128'>
```
May vary.

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -2724,8 +2731,8 @@ def prod(a, axis=None, dtype=None, out=None, keepdims=np._NoValue, initial=np._N
-    >>> np.prod(x)  # random
-    16
+    >>> np.prod(x)
+    6917529010461212688 # may vary
```
This is in the context of "this is how it behaves on a 32-bit platform". We should put it back, or maybe just stick `dtype=np.int32` here

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -2238,6 +2242,9 @@ def all(a, axis=None, out=None, keepdims=np._NoValue):
     >>> z=np.all([-1, 4, 5], out=o)
+    Traceback (most recent call last):
+        ...
+    ValueError: output parameter for reduction operation logical_and has too many dimensions
```
You should fix this line to not error by setting `o=np.array(False)`

eric-wieser commented on numpy/core/function_base.py:
```diff
@@ -308,24 +314,29 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None):
     Graphical illustration of ``endpoint`` parameter:
 
+    >>> import matplotlib
+    >>> matplotlib.use('agg')
```
Does this still render?

tylerjereddy replied:
I added an example of `agg` backend still working in circleci doc artifacts below

eric-wieser commented on numpy/core/numerictypes.py:
```diff
@@ -319,7 +318,7 @@ def issubclass_(arg1, arg2):
     --------
     >>> np.issubclass_(np.int32, int)
-    True
+    False
```
Can you add `# True on python 2.7`

eric-wieser commented on numpy/core/numerictypes.py:
```diff
@@ -485,9 +484,9 @@ def sctype2char(sctype):
     Examples
     --------
-    >>> for sctype in [np.int32, float, complex, np.string_, np.ndarray]:
+    >>> for sctype in [np.int32, np.double, complex, np.string_, np.ndarray]:
```
May as well change `complex` to `np.complex`

eric-wieser commented on numpy/core/numerictypes.py:
```diff
@@ -485,9 +484,9 @@ def sctype2char(sctype):
+    >>> for sctype in [np.int32, np.double, complex, np.string_, np.ndarray]:
     ...     print(np.sctype2char(sctype))
-    l
+    i
```
This is platform-dependent, since int32 is the first of `long` and `int` that's the right size. Either use `np.int_` above, and `l` here, or just mark this as may vary.

charris commented on numpy/fft/fftpack.py:
```diff
@@ -543,16 +543,16 @@ def hfft(a, n=None, axis=-1, norm=None):
     >>> signal = np.array([1, 2, 3, 4, 3, 2])
     >>> np.fft.fft(signal)
-    array([ 15.+0.j,  -4.+0.j,   0.+0.j,  -1.-0.j,   0.+0.j,  -4.+0.j])
+    array([15.+0.j,  -4.+0.j,   0.+0.j,  -1.+0.j,   0.+0.j,  -4.+0.j])
```
`+0.j`, `-0.j` may vary.

charris replied:
And so on below.

eric-wieser commented on numpy/core/records.py:
```diff
@@ -130,18 +129,17 @@ class format_parser(object):
     ...                  ['T1', 'T2', 'T3']).dtype
-    dtype([(('T1', 'col1'), '<f8'), (('T2', 'col2'), '<i4'),
-           (('T3', 'col3'), '|S5')])
+    dtype([(('T1', 'col1'), '<f8'), (('T2', 'col2'), '<i4'), (('T3', 'col3'), 'S5')])
```
This assumes system endianness. Line 130 needs the `<`s too.

eric-wieser commented on numpy/core/records.py:
```diff
@@ -130,18 +129,17 @@ class format_parser(object):
     ...                  []).dtype
-    dtype([('col1', '<f8'), ('col2', '<i4'), ('col3', '|S5')])
+    dtype([('col1', '<f8'), ('col2', '<i4'), ('col3', 'S5')])
     >>> np.format_parser(['f8', 'i4', 'a5'], [], []).dtype
```
`<` needed here and on line 140

eric-wieser commented on numpy/core/records.py:
```diff
@@ -380,20 +378,19 @@ class recarray(ndarray):
     >>> x
-    array([(1.0, 2), (3.0, 4)],
-          dtype=[('x', '<f8'), ('y', '<i4')])
+    array([(1., 2), (3., 4)], dtype=[('x', '<f8'), ('y', '<i8')])
```
Endian problems here again

eric-wieser commented on numpy/core/shape_base.py:
```diff
@@ -179,12 +179,12 @@ def atleast_3d(*arys):
+       [2]]] (1, 2, 1)
     [[[1]
-      [2]]] (1, 2, 1)
+       [2]]] (1, 2, 1)
```
Something weird is going on with the spacing here. Why the doctest skip above?

tylerjereddy replied:
I fixed the spacing but we can't `print()` ndarrays with refguide at the moment -- so, in obvious cases where I can just use the repr I remove the `print()`, otherwise I just add a skip directive.

charris commented on numpy/lib/_iotools.py:
```diff
@@ -146,11 +146,9 @@ def flatten_dtype(ndtype, flatten_base=False):
-    [dtype('|S4'), dtype('float64'), dtype('float64'), dtype('int32'),
-     dtype('int32'), dtype('int32'), dtype('int32'), dtype('int32'),
-     dtype('int32')]
+    [dtype('S4'), dtype('float64'), dtype('float64'), dtype('int64'), dtype('int64'), dtype('int64'), dtype('int64'), dtype('int64'), dtype('int64')]
```
Does this line break?

tylerjereddy replied:
Yes, I've fixed that, though refguide is not as strict about that kind of thing.

eric-wieser commented on numpy/lib/_datasource.py:
```diff
@@ -20,17 +20,18 @@
 Example::
 
     >>> # Create a DataSource, use os.curdir (default) for local storage.
+    >>> import numpy.lib._datasource as datasource
```
This encourages users to use the private module. Can you import it using the public name?

eric-wieser commented on numpy/lib/arrayterator.py:
```diff
@@ -80,9 +80,8 @@ class Arrayterator(object):
-    ...
-    [[[[0 1]]]] (1, 1, 1, 2)
+    ...         print(subarr, subarr.shape) # doctest: +SKIP
+    >>> # [[[[0 1]]]] (1, 1, 1, 2)
```
Why are we skipping this?

tylerjereddy replied:
Again, we can't `print()` ndarrays with refguide -- it will expect `random` for whatever reason--either we use the repr or add the skip directive. Maybe @ev-br knows more.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -2604,15 +2609,15 @@ def blackman(M):
     [<matplotlib.lines.Line2D object at 0x...>]
     >>> plt.title("Blackman window")
-    <matplotlib.text.Text object at 0x...>
+    Text(0.5, 1.0, 'Blackman window')
```
This probably depends on the matplotlib version

tylerjereddy replied:
I guess my approach has been to update to the expectation of a more recent version. Quite a few of these in the diff now.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -3180,11 +3192,14 @@ def kaiser(M, beta):
-             2.29737120e-01,   5.99885316e-01,   9.45674898e-01,
-             9.45674898e-01,   5.99885316e-01,   2.29737120e-01,
-             4.65200189e-02,   3.46009194e-03,   7.72686684e-06])
+     array([7.72686684e-06,   3.46009194e-03,   4.65200189e-02, # may vary
```
extra space on this first line

eric-wieser commented on numpy/lib/nanfunctions.py:
```diff
@@ -1466,11 +1469,11 @@ def nanvar(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue):
     >>> a = np.array([[1, np.nan], [3, 4]])
     >>> np.var(a)
-    1.5555555555555554
+    nan
```
I think the line above should say `nanvar`

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -168,13 +168,13 @@ class NpzFile(Mapping):
     True
     >>> npz.files
-    ['y', 'x']
+    ['x', 'y']
```
I suspect this may be a random ordering. Perhaps add `sorted'?

eric-wieser commented on numpy/lib/npyio.py:
```diff
@@ -597,21 +597,21 @@ def savez(file, *args, **kwds):
     >>> npzfile = np.load(outfile)
     >>> npzfile.files
-    ['y', 'x']
+    ['x', 'y']
```
ordering again

eric-wieser commented on numpy/lib/recfunctions.py:
```diff
@@ -153,9 +157,13 @@ def get_names_flat(adtype):
-    True
+    Traceback (most recent call last):
+        ...
+    AttributeError: 'numpy.ndarray' object has no attribute 'names'
```
This looks like a bug - perhaps just open it later

tylerjereddy replied:
see #12485 -- I'm inclined to leave the "truth" (a traceback) until this is fixed / cleared up, but can add a skip otherwise.

eric-wieser commented on numpy/lib/utils.py:
```diff
@@ -151,9 +151,6 @@ def deprecate(*args, **kwargs):
     >>> olduint(6)
-    /usr/lib/python2.5/site-packages/numpy/lib/utils.py:114:
-    DeprecationWarning: uint32 is deprecated
-      warnings.warn(str1, DeprecationWarning, stacklevel=2)
```
This doesn't look right - the point of this example is to show the deprecation warning

tylerjereddy replied:
I restored the basic deprecation warning, but added a variable flag because:
1) Refguide seems to capture / suppress the warning internally
2) We're deprecating an unsized type so portability isn't great anway

eric-wieser commented on numpy/lib/utils.py:
```diff
@@ -202,7 +199,7 @@ def byte_bounds(a):
     True
     >>> I = np.eye(2, dtype='G'); I.dtype
-    dtype('complex192')
+    dtype('complex256')
```
Don't use `G` here, it's platform-dependent

charris commented on numpy/lib/mixins.py:
```diff
@@ -81,58 +81,55 @@ class NDArrayOperatorsMixin(object):
     class that simply wraps a NumPy array and ensures that the result of any
     arithmetic operation is also an ``ArrayLike`` object::
 
-        class ArrayLike(np.lib.mixins.NDArrayOperatorsMixin):
```
@shoyer would prefer without `...` prefixes  below. Needs some experimentation.

eric-wieser commented on runtests.py:
```diff
@@ -202,6 +202,14 @@ def main(argv):
+               '--doctests']
+        if args.submodule:
+            cmd += [args.submodule]
+        os.execv(sys.executable, [sys.executable] + cmd)
```
Probably better to use subprocess

tylerjereddy replied:
This changes the behavior slightly in my hands when using `subprocess.call([sys.executable] + cmd)`. In particular, the output text includes a few lines about building that I don't see in the original output, but otherwise seems fine.

eric-wieser commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -2949,25 +2953,25 @@ cdef class RandomState:
         >>> plt.figure()
         >>> plt.hist(1./(1.+rvsp), bins=50, density=True)
-        >>> plt.plot(xx,powpdf,'r-')
+        >>> plt.plot(xx,powpdf,'r-') 
```
Lots ot trailing spaces added here

tylerjereddy committed:
MAINT: address several reviewer comments

shoyer commented:
With regards to executing doctests on traditional code blocks, it looks like the Manuel package was developed specifically for this:
https://pypi.org/project/manuel/
https://pythonhosted.org/manuel/

Is this a reasonable dependency to add to our build?

It would be a nice way to handle doctests that include code blocks like `NDArrayOperatorsMixin`.

tylerjereddy commented:
Maybe it makes sense to add some third-party solutions like that for fine-tuning / appearance, but I'd say I have my hands full just handling the vanilla refguide check work & plethora of reviewer comments here for now. Maybe for follow-up PRs, etc.?

And now the burden of the matmul docstrings is on me too -- merge commits are finding issues there & will continue to find new issues for anything that adds / adjusts new docstrings moving forward unless they are carefully checked immediately before merging.

tylerjereddy committed:
MAINT: matmul docstring marked with SKIP in some places until fixed

tylerjereddy committed:
MAINT: clean up whitespace issues in mtrand.pyx.

tylerjereddy committed:
MAINT: revert numpy/lib/mixins to master branch status -- refguide passes in any case now.

tylerjereddy commented:
Don't even need the SKIP directive for `NDArrayOperatorsMixin` example -- I reverted it to master branch status without prompts and refguide is just fine with it (well, assuming it is actually executed). I opened a connected issue for any further discussion on the matter of prompts in docs.

tylerjereddy commented:
My notes at the moment show:

- spot checks for matplotlib -- are the plots still being produced in cases where i.e., I switched backend to `agg` or similar?
- fixed-size dtype changes -- quite a few of these cases to be careful with?
- subprocess -- worth changing refguide vendored in from SciPy for this?
- `>>> #` comment changes should perhaps be reverted to complete absence of prompt, though maybe not for cases that I didn't change myself

charris commented:
OK, thanks for the update.

eric-wieser commented on numpy/core/defchararray.py:
```diff
@@ -1090,7 +1090,7 @@ def lstrip(a, chars=None):
     array(['aAaAaA', '  aA  ', 'abBABba'], dtype='<U7')
     >>> (np.char.lstrip(c, ' ') == np.char.lstrip(c, '')).all()
-    ... # XXX: is this a regression? this line now returns False
+    ... # This used to return True
```
Maybe keep `XXX: is this a regression? This used to return True`

tylerjereddy replied:
alright, done

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -2240,13 +2240,10 @@ def all(a, axis=None, out=None, keepdims=np._NoValue):
-    >>> id(z), id(o), z                             # doctest: +SKIP
-    (28293632, 28293632, array([ True]))
+    >>> id(z), id(o), z
+    (28293632, 28293632, array([ True])) # may vary
```
Is `z` really a 1d array here?

tylerjereddy replied:
No, it probably hasn't been tested in a long time because of the SKIP directive (and years without doctests, etc.) -- I've adjusted accordingly, though the proper result still won't be protected from regression with the in-line variable tag on the memory addresses.

Maybe better long-term to i.e., split on two separate lines, but I'll mark as resolved -- micro-improvement can always be made on a case-by-case basis in follow-ups.

eric-wieser commented on numpy/lib/utils.py:
```diff
@@ -198,8 +198,8 @@ def byte_bounds(a):
-    >>> I = np.eye(2, dtype='G'); I.dtype
-    dtype('complex256')
+    >>> I = np.eye(2); I.dtype
+    dtype('complex256') # may vary
```
No longer true - since you removed the `dtype='G'`, this will always output float64.

tylerjereddy replied:
revised accordingly

eric-wieser commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -2605,9 +2605,9 @@ def add_newdoc(place, name, doc):
-      >>> np.dot(a, c).shape
+      >>> np.dot(a, c).shape # doctest: +SKIP
       (9, 5, 7, 9, 5, 4)
-      >>> np.matmul(a, c).shape # n is 5, k is 3, m is 4
```
Is there an easy way to keep the original comment too? Maybe move it to the next line?

tylerjereddy replied:
I restored the comment below the original output while retaining the new SKIP directive until fixed by someone later.

tylerjereddy committed:
MAINT: addressing review comments

* restored regression comment in
numpy/core/defchararray.py

* fixed the dimensionality of the z
array in all() docstring in
numpy/core/fromnumeric.py; this isn't
detected because it is in-line with
variable memory addresses which are
tagged as variable for refguide

* byte_bounds() docstring adjusted
to reflect non-variable dtype
after reviewer requested removal
of complex dtype

* restore an original comment
in matmul docstring, as requested
by reviewer

tylerjereddy committed:
MAINT: >>> # style cleanups requested

* reviewer requested that the cases where
I switched from free-floating comments to
`>>> # comments` be reverted to free-floating
in docstrings

tylerjereddy commented:
Here's an [example circleci doc build artifact](https://102-116427166-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/reference/generated/numpy.linspace.html?highlight=linspace#numpy.linspace) showing that a plot is still produced with my backend switching adjustments. That stuff may be better abstracted somewhere else so as to not distract, but maybe ok for now.

tylerjereddy commented:
`numpy.core.tests.test_multiarray.TestMatmul` stochastic failure on 32-bit?

charris commented:
That's a strange failure. Hmm.... I've restarted the azure builds.

charris commented:
What happens if the tests encounter an invalid value in the doctests? In the "official" tests we suppress warnings when they don't matter.

tylerjereddy committed:
MAINT: include recfunctions module

* added lib.recfunctions to refguide
PUBLIC_SUBMODULES, as the doctests
were otherwise not getting executed

* fixed a failing doctest in
recfunctions after above activation

charris commented:
I'm a bit wary about trying to get this into 1.16, so giving it a 1.17 milestone so we have time to see and deal with any issues that may arise. It would probably be OK, maybe even desirable, to put in the docstring fixes and just leave the testing turned off until 1.17. Thoughts?

tylerjereddy commented on tools/refguide_check.py:
```diff
@@ -0,0 +1,952 @@
+
+BASE_MODULE = "numpy"
+
+PUBLIC_SUBMODULES = [
```
It is important to check that this list of public NumPy submodules is reasonably complete. I just noticed that `lib.recfunctions` was previously absent, so refguide_check wasn't executing docstrings in that module, some of which were the subject of reviewer comments (probably diffs originated from before refguide switch when I was using vanilla doctests).

Also, even if I add `lib.recfunctions`, not all function docstrings therein are executed -- it seems I also need to add functions to `__all__` in the module, but I'm hesitant do start doing that more broadly without reviewer approval first. Thoughts on this?

charris replied:
I think skipping private functions, which should be everything not in `__all__` with, or without the underscore, would probably be OK.  If they are not intended to be private then they should be added. That said, `lib.recfunctions` has always been an oddball, as it is not imported at the top level.

@shoyer Not all functions in `lib.recfunctions` have dispatchers.

shoyer replied:
@charris looking through `lib.recfunctions` the functions without dispatchers seem to not take arrays as inputs (e.g., `get_fieldspec`). So these properly don't have dispatchers -- there's no array argument on which to call a dispatcher method (we don't have duck-typed dtypes). Let me know if there's anything in particular that you think I missed, though...

charris commented:
I was thinking of merging as is, but with the doctests turned off. You could then open another PR, turn them on, and continue making fixes. The advantage would be similarity between 1.16 and 1.17, the disadvantage would be loss of all the comments, which might be too high a price to pay.

tylerjereddy commented on numpy/lib/mixins.py:
```diff
@@ -81,58 +81,55 @@ class NDArrayOperatorsMixin(object):
+    ...        return '%s(%r)' % (type(self).__name__, self.value)
+
+    >>> # In interactions between ``ArrayLike`` objects and numbers or numpy arrays,
+    >>> # the result is always another ``ArrayLike``:
```
I think you're looking at an outdated diff -- there's even a badge above indicating as such.

I checked out numpy/lib/mixins from latest master so there shouldn't be a diff from your original code. That said, doctests probably aren't even running for lib.mixins until someone, probably me, adds it to the list of public modules in refguide.

tylerjereddy commented on numpy/lib/mixins.py:
```diff
@@ -81,58 +81,55 @@ class NDArrayOperatorsMixin(object):
+    >>> import numbers
+    >>> class ArrayLike(np.lib.mixins.NDArrayOperatorsMixin):
+    ...    def __init__(self, value):
+    ...        self.value = np.asarray(value)
```
The code was reverted to master branch status already based on your previous comments.

shoyer replied:
Oops -- sorry about these outdated comments! I had drafted them before but not sent them -- and they just got sent along with my latest response to Chunk!

charris commented:
> So these properly don't have dispatchers -- 

Oh, my bad...

mattip commented:
At some point [`_pytesttester.py`](https://github.com/numpy/numpy/blob/maintenance/1.16.x/numpy/_pytesttester.py#L175) needs to be updated to pass the `doctests` option through

eric-wieser reviewed (approved):
I think there's enough good stuff here that we should go ahead and merge - this diff is too large to catch everything in, and even if there are a few points of disagreement, the docstrings are overall way more accurate.

The cost will at worst be intermittent CI failures, which we can either catch quickly, or resolve temporarily by disabling doctests again while we sort them out.

tylerjereddy commented:
Ok, I rebased and force-pushed. If the CI remains all-green I suppose we could go for merge if others are ok with a sort of "2 steps forward, 1 step backward" approach, with smaller / focused PRs to follow.

I think @stefanv wants a "wish list" for refguide features/ fixes. There has been some discussion of eventually extracting to its own package for the community, but one thing at a time.

The current apparent opt-in instead of opt-out model seems a little annoying. On that note, I do see a few posts floating around online about combining coverage with doctests, one from [Raymond Hettinger](https://stackoverflow.com/a/45262687/2942522), and another [pytest issue](https://github.com/pytest-dev/pytest-cov/issues/101) apparently indicating that this does actually work for doctests + coverage, but people aren't sure how the magic works.

charris commented:
Thanks Tyler.
